drivers: Add support for pinctrl in nRF SPI and I2C drivers

anangl opened the pull request:
Add support for the new pinctrl API to the SPI and I2C drivers that handle the following nRF peripherals: SPI, SPIM, SPIS, TWI, and TWIM.
Add also build assertions in the nRF UART drivers that will ensure that every peripheral for which a driver instance is created has some pins assigned to it.

anangl committed:
drivers: serial: nrfx: Ensure that instances have some pins assigned

Add build assertions that will ensure that every peripheral for
which a driver instance is created has some pins assigned to it.
Neither pinctrl-0 nor *-pin properties can be currently marked as
required in devicetree, so these assertions will help users avoid
invalid configurations where it could be hard to figure out why
the UART is not working.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pinctrl: nrf: Add support for SPI and TWI peripherals

Add support for configuring pins of the following nRF peripherals:
SPI, SPIM, SPIS, TWI, and TWIM.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: spi: nrfx: Clean up code

- clean up registration of the drivers with the logging subsystem
- use consistent naming of local variables accessing configuration
  and runtime data of driver instances, for easier code maintenance

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: spi: nrfx: Add support for pinctrl

Add support for the new pinctrl API to the SPI drivers that handle
the nRF SPI, SPIM, and SPIS peripherals. Update code of the drivers
and related devicetree bindings.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: i2c: nrfx: Add support for pinctrl

Add support for the new pinctrl API to the I2C drivers that handle
the nRF TWI and TWIM peripherals. Update code of the drivers and
related devicetree bindings.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -1192,6 +1192,13 @@ static int uart_nrfx_pm_action(const struct device *dev,
 #endif /* CONFIG_PINCTRL */
 
+BUILD_ASSERT((IS_ENABLED(CONFIG_PINCTRL) &&
+	      DT_PINCTRL_HAS_IDX(DT_DRV_INST(0), 0)) ||
```
entry "sleep" is also required if `CONFIG_PM_DEVICE=y`.

gmarull commented on samples/boards/nrf/dynamic_pinctrl/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -17,6 +17,11 @@
 	pinctrl-names = "default", "sleep";
 };
 
+/* Disable uart1 so that it is not left without any pins assigned. */
+&uart1 {
+	status = "disabled";
+};
```
Please add a TODO here about removing this once all boards are ported

gmarull commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -277,8 +291,13 @@ static int spi_nrfx_pm_action(const struct device *dev,
 	case PM_DEVICE_ACTION_RESUME:
-		/* No action needed at this point, nrfx_spi_init() will be
-		 * called at configuration before the next transfer.
+		ret = spi_nrfx_pins_init(dev);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* nrfx_spi_init() will be called at configuration before
+		 * the next transfer.
```
why do who have a helper here? I'd be consistent and use #ifdefs everywhere, as in the sleep case.

anangl replied:
Default configuration of the pins needs to be also applied when the driver instance is initialized (in `spi_##idx##_init()`). That's why `spi_nrfx_pins_init()` was introduced.

gmarull replied:
why can't we use regular ifdeffed pinctrl calls? `spi_nrfx_pins_init` naming is confusing IMO, and inconsistent with the sleep counterpart.

anangl replied:
I was afraid that `spi_##idx##_init()`, which is in a macro, would be difficult to read then. But I agree, it is confusing it the current form. I'll get rid of `spi_nrfx_pins_init()`.

gmarull commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -317,9 +344,24 @@ static int spi_nrfx_pm_action(const struct device *dev,
 #define SPI_NRFX_SPI_DEVICE(idx)					       \
-	BUILD_ASSERT(							       \
-		!SPI_PROP(idx, miso_pull_up) || !SPI_PROP(idx, miso_pull_down),\
+	BUILD_ASSERT((IS_ENABLED(CONFIG_PINCTRL) &&			       \
+		      DT_PINCTRL_HAS_IDX(SPI(idx), 0)) ||		       \
+		     (!IS_ENABLED(CONFIG_PINCTRL) &&			       \
+		      DT_NODE_HAS_PROP(SPI(idx), sck_pin)),		       \
+		"No pins assigned to SPI"#idx);				       \
```
ditto, sleep case should also be checked. I think creating an intermediate helper macro would improve readability here

anangl replied:
I'll add a generic macro for checking this, to not copy almost the same code in several drivers.

anangl replied:
Added.

gmarull commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -435,8 +450,13 @@ static int spim_nrfx_pm_action(const struct device *dev,
 	case PM_DEVICE_ACTION_RESUME:
-		/* No action needed at this point, nrfx_spim_init() will be
-		 * called at configuration before the next transfer.
+		ret = spi_nrfx_pins_init(dev);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* nrfx_spim_init() will be called at configuration before
+		 * the next transfer.
```
ditto

gmarull commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -228,6 +243,11 @@ static int twi_nrfx_pm_action(const struct device *dev,
 
 	switch (action) {
 	case PM_DEVICE_ACTION_RESUME:
+		ret = twim_nrfx_pins_init(dev);
+		if (ret < 0) {
+			return ret;
+		}
```
ditto

anangl commented:
Rebased.

anangl commented:
The compliance check fails because of this macro call:
```
		      (FOR_EACH_FIXED_ARG(NRF_DT_ENSURE_NODE_HAS_PROP,	\
					  (||), node_id, __VA_ARGS__))),\
```
but I see no reasonable way of fixing this. The `||` operator must be in parentheses because the macro requires this. For the above form, checkpatch signals an error because of "spaces required around that '||'". For `( || )` in turn, it signals errors because of prohibited spaces after the opening parenthesis and before the closing one...
This could be fixed with:
```
		      (FOR_EACH_FIXED_ARG(NRF_DT_ENSURE_NODE_HAS_PROP, (\
					  ||), node_id, __VA_ARGS__))), \
```
but then the code looks weird and can be confusing, so I'm not sure if that's worth it.

gmarull commented:
@anangl maybe we can patch checkpatch with something like:

```diff
diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index fb2efb1d09..cc702c91d0 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -4726,6 +4726,13 @@ sub process {
                                                $ok = 1;
                                        }
 
+                                       # some macros require a separator argument to be in parenthesis,
+                                       # e.g. (||).
+                                       if (($op eq '||') &&
+                                           ($ca =~ /\($/ || $cc =~ /^\)/)) {
+                                               $ok = 1;
+                                       }
+
                                        # messages are ERROR, but ?: are CHK
                                        if ($ok == 0) {
                                                my $msg_level = \&ERROR;
```

anangl commented:
> @anangl maybe we can patch checkpatch with something like:

@gmarull If it is allowed to modify checkpatch this way, that works for me.
Would you mind creating a commit with this patch so that I could include it in the PR, or should I create one with you as the author (as this piece of code is entirely yours)?

gmarull committed:
scripts: checkpatch: allow parentheses around operators

Some macros, e.g., FOR_EACH_FIXED_ARG, require a separator argument that
needs to be in parentheses, e.g., (||). This should not trigger a
checkpatch warning.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
